src/app-layer-ssl.c

```diff
@@ -834,37 +844,49 @@ static inline int TLSDecodeHSHelloCipherSuites(SSLState *ssl_state,
         goto invalid_length;
     }
 
-    if (SC_ATOMIC_GET(ssl_config.enable_ja3)) {
-        JA3Buffer *ja3_cipher_suites = Ja3BufferInit();
-        if (ja3_cipher_suites == NULL)
-            return -1;
+    if (SC_ATOMIC_GET(ssl_config.enable_ja3) || SC_ATOMIC_GET(ssl_config.enable_ja4)) {
+        JA3Buffer *ja3_cipher_suites = NULL;
+
+        if (SC_ATOMIC_GET(ssl_config.enable_ja3)) {
+            ja3_cipher_suites = Ja3BufferInit();
+            if (ja3_cipher_suites == NULL)
+                return -1;
+        }
 
         uint16_t processed_len = 0;
         /* coverity[tainted_data] */
         while (processed_len < cipher_suites_length)
         {
             if (!(HAS_SPACE(2))) {
-                Ja3BufferFree(&ja3_cipher_suites);
+                if (SC_ATOMIC_GET(ssl_config.enable_ja3)) {
```

Comment:
is there a race condition here? If `ssl_config.enable_ja3` flips to `false` after the alloc, but before this check, then we'd not free it? I wonder if it makes more sense to init it to a const
`const bool enable_ja3 = SC_ATOMIC_GET(ssl_config.enable_ja3);`

Reply:
alternatively, just check the pointer `ja3_cipher_suites`. If it is non-NULL, free it regardless?

Reply:
for the logic below, the ptr check could also work I think

Reply:
`ssl_config.enable_ja3` is only set in `RegisterSSLParsers()` (which I assume is only called once during the initial setup) and `SSLEnableJA3()` which is only called during rule keyword setup (to enable the feature if a rule uses it). I would expect any of these calls to return before any parsing ever takes place.

But it doesn't hurt to store the value at the beginning of `TLSDecodeHSHelloCipherSuites()` in a local `const` value to ensure it definitely stays consistent across the whole runtime of the function. I'd be happy to make this change.

Reply:
If that is all you have, I'd just push one more PR.